async_generator/_tests/test_async_generator.py

```diff
@@ -601,31 +619,83 @@ def __init__(self, count):
 ################################################################
 
 
-async def test___del__():
-    gen = async_range(10)
+async def test___del__(capfd):
+    completions = 0
+
+    @async_generator
+    async def awaits_when_unwinding():
+        await yield_(1)
+        try:
+            await yield_(2)
+        finally:
+            await mock_sleep()
+        try:
+            await yield_(3)
+        finally:
+            nonlocal completions
+            completions += 1
+
+    gen = awaits_when_unwinding()
     # Hasn't started yet, so no problem
     gen.__del__()
 
-    gen = async_range(10)
-    await collect(gen)
+    gen = awaits_when_unwinding()
+    assert await collect(gen) == [1, 2, 3]
     # Exhausted, so no problem
     gen.__del__()
 
-    gen = async_range(10)
-    await gen.aclose()
-    # Closed, so no problem
-    gen.__del__()
+    for turns in (1, 2, 3):
+        gen = awaits_when_unwinding()
+        for turn in range(1, turns + 1):
```

Comment:
If you make it yield `0, 1, 2` then this will be a bit simpler :-)

Reply:
done